File: src/main/scala/com/digitalasset/testing/ledger/clock/FixedTimeProvider.scala
/*
 * Copyright 2019 Digital Asset (Switzerland) GmbH and/or its affiliates
 *
 * SPDX-License-Identifier: Apache-2.0
 */

package com.digitalasset.testing.ledger.clock

import java.time.Instant

import io.grpc.ManagedChannel

class FixedTimeProviderFactory(private val fixedTime: Instant)
    extends TimeProviderFactory {
  override def getTimeProvider(channel: ManagedChannel,
                               ledgerId: String): TimeProvider =
    new FixedTimeProvider(fixedTime)
}

class FixedTimeProvider(private var fixedTime: Instant) extends TimeProvider {
  override def getCurrentTime: Instant = fixedTime

  override def setCurrentTime(time: Instant): Unit = fixedTime = time
}

File: src/main/java/com/digitalasset/testing/ledger/clock/FixedTimeProvider.java
/*
 * Copyright 2019 Digital Asset (Switzerland) GmbH and/or its affiliates
 *
 * SPDX-License-Identifier: Apache-2.0
 */

package com.digitalasset.testing.ledger.clock;

import java.time.Instant;
import java.util.function.Supplier;

public class FixedTimeProvider implements TimeProvider {
  public static Supplier<TimeProvider> factory(Instant fixedTime) {
    return () -> new FixedTimeProvider(fixedTime);
  }

  private volatile Instant fixedTime;

  public FixedTimeProvider(Instant fixedTime) {
    this.fixedTime = fixedTime;
  }

  @Override
  public Instant getCurrentTime() {
    return fixedTime;
  }

  @Override
  public void setCurrentTime(Instant time) {
    fixedTime = time;
  }
}
